apps/voting-disputable/contracts/DisputableVoting.sol

```diff
@@ -0,0 +1,803 @@
+        return _newVote(_executionScript, _context);
+    }
+
+    /**
+    * @notice Vote `_supports ? 'yes' : 'no'` in vote #`_voteId`
+    * @dev Initialization check is implicitly provided by `voteExists()` as new votes can only be
+    *      created via `newVote()`, which requires initialization
+    * @param _voteId Id for vote
+    * @param _supports Whether voter supports the vote
+    */
+    function vote(uint256 _voteId, bool _supports) external voteExists(_voteId) {
+        require(_canVote(votes[_voteId], msg.sender), ERROR_CANNOT_VOTE);
+        _vote(_voteId, _supports, msg.sender);
+    }
+
+    /**
+    * @notice Vote `_supports ? 'yes' : 'no'` in vote #`_voteId` on behalf of multiple voters
+    * @dev Initialization check is implicitly provided by `voteExists()` as new votes can only be
+    *      created via `newVote()`, which requires initialization
+    * @param _voteId Id for vote
+    * @param _supports Whether the representative supports the vote
+    * @param _voters Addresses of the voters to vote on behalf of
+    */
+    function voteOnBehalfOf(uint256 _voteId, bool _supports, address[] _voters) external voteExists(_voteId) {
+        require(_voters.length <= MAX_VOTES_DELEGATION_SET_LENGTH, ERROR_DELEGATES_EXCEEDS_MAX_LEN);
+        _voteOnBehalfOf(_voteId, _supports, _voters);
+    }
+
+    /**
+    * @notice Execute vote #`_voteId`
+    * @dev Initialization check is implicitly provided by `voteExists()` as new votes can only be
+    *      created via `newVote()`, which requires initialization
+    * @param _voteId Id for vote
+    */
+    function executeVote(uint256 _voteId) external voteExists(_voteId) {
+        Vote storage vote_ = votes[_voteId];
+        require(_canExecute(vote_), ERROR_CANNOT_EXECUTE);
+
+        vote_.status = VoteStatus.Executed;
+        // TODO: check agreement action is not closed
+        _closeAgreementAction(vote_.actionId);
+
+        // Add Agreement to blacklist to disallow the stored EVMScript from directly calling the
+        // linked Agreement from this app's context (e.g. maliciously creating a new action)
+        address[] memory blacklist = new address[](1);
+        blacklist[0] = address(_getAgreement());
+        runScript(vote_.executionScript, new bytes(0), blacklist);
+        emit ExecuteVote(_voteId);
+    }
+
+    // Disputable getter fns
+
+    /**
+    * @dev Tells whether vote #`_voteId` can be challenged
+    * @param _voteId Id for vote
+    * @return True if the given vote can be challenged
+    */
+    function canChallenge(uint256 _voteId) external view voteExists(_voteId) returns (bool) {
+        return _isVoteOpen(votes[_voteId]);
+    }
+
+    /**
+    * @dev Tells whether vote #`_voteId` can be closed
+    * @param _voteId Id for vote
+    * @return True if the given vote can be closed
+    */
+    function canClose(uint256 _voteId) external view voteExists(_voteId) returns (bool) {
+         return !_isVoteOpen(votes[_voteId]);
```

Comment:
Want to double check this case—right now, this looks like you can close a vote once you've challenged it.

While, I believe, the Agreement does not let you do this from `closeAction()` due to the internal `_canProceed()` check, we may still want to be a bit more explicit here.

With the previous comment about making `closeAction()` potentially a bit smarter when it's called from the disputable app directly, this change shouldn't significantly impact gas in the happy path (executing a passed vote).

Reply:
Yes, that's true. However, I prefer making this explicit, let's change it to state that votes can only be closed if they are still active and the vote end date has passed.